### PR TITLE
Add wallet balance transfer Page 

### PR DESCRIPTION
**Pull request by parasharrajat**
Please review @marcaaron 

### Fixed Issues

$ https://github.com/Expensify/App/issues/3922

### Tests | QA Steps
1. Open Payments page.
2. Click on the wallet transfer.
3. You should be taken to the Transfer balance page.
4. If your wallet balance is greater than the fee and which account? has an account then you should be able to click the transfer button and it should be enabled.
5. Observe if **Which account?** is a debit card then the Instant method is selected.
6. Observe if **Which account?** is a bank account then the 1-3 business days method is selected.
7. Paypal should not be shown as **Which account?**

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web | Mobile Web | Desktop

image

### CODE REVIEW

File: src/libs/actions/PaymentMethods.js [hunk 1]
```diff
@@ -95,8 +96,70 @@ function clearDebitCardFormErrorAndSubmit() {
...
+    Onyx.merge(ONYXKEYS.WALLET_TRANSFER, data);
+}
+
+/**
+ * Cancel the wallet transfer
+ */
+function cancelWalletTransfer() {
+    Onyx.set(ONYXKEYS.WALLET_TRANSFER, null);
+}
+
```
Comment (parasharrajat) [inline src/libs/actions/PaymentMethods.js:156]: I am not sure what would be the best way to handle these but I need these to share data across screens.

File: src/pages/settings/Payments/TransferBalancePage.js [hunk 2]
```diff
@@ -1,27 +1,246 @@
...
+        );
+        const selectedAccount = this.props.walletTransfer.selectedAccountID
+            ? _.find(
+                paymentMethods,
+                method => method.id === this.props.walletTransfer.selectedAccountID,
+            )
+            : defaultAccount;
+
+        return selectedAccount;
+    }
```
Comment (parasharrajat) [inline src/pages/settings/Payments/TransferBalancePage.js:109]: I don't see any selected account by default. `userWallet.walletLinkedAccountID` does not have any value.

**Comment by parasharrajat** [comment]
Note: I can't test the wallet transfer, I would need your help to do that. Thanks.

File: src/pages/settings/Payments/TransferBalancePage.js [hunk 3]
```diff
@@ -1,27 +1,246 @@
...
+            },
+        ];
+
+        this.transferBalance = this.transferBalance.bind(this);
+
+        const selectedAccount = this.getSelectedAccount();
+        PaymentMethods.saveWalletTransferAmountAndAccount(
+            this.props.userWallet.currentBalance,
+            selectedAccount ? selectedAccount.id : '',
+        );
```
Comment (marcaaron) [inline src/pages/settings/Payments/TransferBalancePage.js:84]: Does it make sense to do this in the `constructor()`? If the `currentBalance` changes the user will transfer whatever we first had when the page opened and not the new balance.

File: src/pages/settings/Payments/TransferBalancePage.js [hunk 4]
```diff
@@ -1,27 +1,246 @@
...
+    }
+
+    /**
+     * Get the selected/default Account for wallet tsransfer
```
Comment (marcaaron) [inline src/pages/settings/Payments/TransferBalancePage.js:88]:
```suggestion
     * Get the selected/default payment method account for wallet transfer
```

File: src/pages/settings/Payments/TransferBalancePage.js [hunk 5]
```diff
@@ -1,27 +1,246 @@
...
+     * @returns {Object|undefined}
+     */
+    getSelectedAccount() {
```
Comment (marcaaron) [inline src/pages/settings/Payments/TransferBalancePage.js:91]:
```suggestion
    getSelectedPaymentMethodAccount() {
```

File: src/libs/PaymentUtils.js [hunk 6]
```diff
@@ -10,7 +10,7 @@ import * as Localize from './Localize';
  * @param {Array} bankAccountList
  * @param {Array} cardList
  * @param {String} [payPalMeUsername='']
- * @returns {Array<PaymentMethod>}
+ * @returns {Array<Object>}
  */
 function getPaymentMethods(bankAccountList, cardList, payPalMeUsername = '') {
```
Comment (marcaaron) [inline src/libs/PaymentUtils.js:15]: I find it pretty confusing that we have both

`PaymentMethods.getPaymentMethods()` and `PaymentUtils.getPaymentMethods()`

Probably we should rename this to something like `formatPaymentMethods()` since it doesn't actually "get" them and processes a list of things we pass in.

File: src/libs/actions/PaymentMethods.js [hunk 7]
```diff
@@ -95,8 +96,70 @@ function clearDebitCardFormErrorAndSubmit() {
...
+    };
+    Onyx.merge(ONYXKEYS.WALLET_TRANSFER, {loading: true});
+
+    return API.TransferWalletBalance(parameters)
+        .then((response) => {
+            if (response.jsonCode !== 200) {
+                throw new Error(response.message);
+            }
+            Onyx.merge(ONYXKEYS.USER_WALLET, {balance: 0});
+            Onyx.merge(ONYXKEYS.WALLET_TRANSFER, {completed: true, loading: false});
```
Comment (marcaaron) [inline src/libs/actions/PaymentMethods.js:118]: The presence of `completed` and `loading` confusing. The only place where we use "completed" is to show the "confirm modal" so perhaps it would be better to call this something like `shouldShowConfirmModal: true`.

File: src/pages/settings/Payments/PaymentsPage.js [hunk 8]
```diff
@@ -110,6 +121,13 @@ class PaymentsPage extends React.Component {
         this.setState({shouldShowAddPaymentMenu: false});
     }
 
+    /**
+    * Navigate to Transfer wallet balance page
+    */
```
Comment (marcaaron) [inline src/pages/settings/Payments/PaymentsPage.js:126]: This comment can be removed as it is obvious what happens.

File: src/libs/actions/PaymentMethods.js [hunk 9]
```diff
@@ -95,8 +96,70 @@ function clearDebitCardFormErrorAndSubmit() {
...
+}
+
+/**
+ * Set the necessary data for wallet transfer
+ * @param {Number} currentBalance
+ * @param {Number} selectedAccountID
+ */
+function saveWalletTransferAmountAndAccount(currentBalance, selectedAccountID) {
+    Onyx.set(ONYXKEYS.WALLET_TRANSFER, {
+        transferAmount: currentBalance - CONST.WALLET.TRANSFER_BALANCE_FEE,
```
Comment (marcaaron) [inline src/libs/actions/PaymentMethods.js:134]: If we already have the `userWallet.currentBalance` why do we need to store this data again as "transferAmount"? Seems like the amount to transfer should always be the `currentBalance`?

File: src/pages/settings/Payments/TransferBalancePage.js (hunk 2 above)
Comment (marcaaron) [reply src/pages/settings/Payments/TransferBalancePage.js:109]: I _think_ we have to call the `SetWalletLinkedAccount` command to set one. @stitesExpensify would maybe know for sure.

File: src/pages/settings/Payments/TransferBalancePage.js [hunk 10]
```diff
@@ -1,27 +1,246 @@
...
+            method => method.id === lodashGet(this.props, 'userWallet.walletLinkedAccountID', ''),
+        );
+        const selectedAccount = this.props.walletTransfer.selectedAccountID
+            ? _.find(
+                paymentMethods,
+                method => method.id === this.props.walletTransfer.selectedAccountID,
+            )
+            : defaultAccount;
+
+        return selectedAccount;
```
Comment (marcaaron) [inline src/pages/settings/Payments/TransferBalancePage.js:108]:
```suggestion
        const accountID = this.props.walletTransfer.selectedAccountID || lodashGet(this.props, 'userWallet.walletLinkedAccountID', '');
        return _.find(paymentMethods, method => method.id === accountID);
```

File: src/pages/settings/Payments/TransferBalancePage.js [hunk 11]
```diff
@@ -1,27 +1,246 @@
...
+    }
+
+    /**
+     * Transfer Wallet balance
+     * @param {PaymentMethod} selectedAccount
+     */
+    transferBalance(selectedAccount) {
+        if (!selectedAccount) {
+            return;
+        }
```
Comment (marcaaron) [inline src/pages/settings/Payments/TransferBalancePage.js:118]: Is there some situation where we would expect `selectedAccount` to not exist?

File: src/pages/settings/Payments/paymentPropTypes.js [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+    /** Whether the user has intiatied the tranfer and transfer request is submitted to backend. */
+    completed: PropTypes.bool,
+});
+
+export {
+    bankAccountListPropTypes,
+    cardListPropTypes,
+    walletTransferPropTypes,
+};
```
Comment (marcaaron) [inline src/pages/settings/Payments/paymentPropTypes.js:50]: Most of our propTypes up until this point have not been combined. Not sure if we should start now and think we should have one file per set of propTypes with a matching file name.

File: src/pages/settings/Payments/paymentPropTypes.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
+import PropTypes from 'prop-types';
+import CONST from '../../../CONST';
+
+/** Array of bank account objects */
+const bankAccountListPropTypes = PropTypes.arrayOf(PropTypes.shape({
```
Comment (marcaaron) [inline src/pages/settings/Payments/paymentPropTypes.js:5]: It seems these are only used once and therefore do not need to be moved into their own file.

File: src/pages/settings/Payments/paymentPropTypes.js [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+    /** The bankAccountID in the bankAccounts db */
+    bankAccountID: PropTypes.number,
+
+    /** The bank account type */
+    type: PropTypes.string,
+}));
+
+/** Array of card objects */
+const cardListPropTypes = PropTypes.arrayOf(PropTypes.shape({
```
Comment (marcaaron) [inline src/pages/settings/Payments/paymentPropTypes.js:20]: Also only discovered one usage for these so they do not need to be in their own file.

**Review by marcaaron — CHANGES_REQUESTED** [review]

File: src/pages/settings/Payments/TransferBalancePage.js (hunk 3 above)
Comment (parasharrajat) [reply src/pages/settings/Payments/TransferBalancePage.js:84]: Actually, the balance is transferred to the backend. A just takes thee account Di where the user wants to transfer. But there is no other way on the app where wallet values will update automatically. The wallet is only refreshed when we open some pages and the First-time app loads.

But I agree I can subscribe to wallet wherever I need the transfer amount.

File: src/libs/actions/PaymentMethods.js [hunk 15]
```diff
@@ -95,8 +96,70 @@ function clearDebitCardFormErrorAndSubmit() {
...
+    });
+}
+
+/**
+ * Update selected accountID and other data for wallet transfer
+ * @param {Object} data
+ */
+function updateWalletTransferData(data) {
+    Onyx.merge(ONYXKEYS.WALLET_TRANSFER, data);
+}
```
[the remark this answers is not in the document]
Comment (parasharrajat) [reply src/libs/actions/PaymentMethods.js:148]: not needed yet. I will remove it.

File: src/libs/actions/PaymentMethods.js (hunk 1 above)
Comment (parasharrajat) [reply src/libs/actions/PaymentMethods.js:156]: It was just the shorter way to do two things at once. reset the data and hide the confirm modal.

**Comment by parasharrajat** [comment]
Updating the PR asap

File: src/pages/settings/Payments/TransferBalancePage.js (hunk 3 above)
Comment (parasharrajat) [reply src/pages/settings/Payments/TransferBalancePage.js:84]: I wanted to remove the temporary data that is added for wallet transfer. we can't clear the data(set that to null) before Confirm modal is shown as the same key `ONYXKEYS.WALLET_TRANSFER.shouldShowConfirmModal` would be used to show the modal.

After the modal is shown and use close it. We can remove all the temporary data from Onyx by setting it to null.

@marcaaron

File: src/libs/actions/PaymentMethods.js (hunk 9 above)
Comment (parasharrajat) [reply src/libs/actions/PaymentMethods.js]: I have changed the code to use the `currentBalance` directly from `userWallet` Onyx key.

File: src/pages/settings/Payments/TransferBalancePage.js (hunk 3 above)
Comment (marcaaron) [reply src/pages/settings/Payments/TransferBalancePage.js]: Sorry, I don't understand the explanation. Can you try asking this in the form of a question or point to some examples?

File: src/libs/PaymentUtils.js [hunk 16]
```diff
@@ -64,6 +64,16 @@ function getPaymentMethods(bankAccountList, cardList, payPalMeUsername = '') {
     return combinedPaymentMethods;
 }
 
+/**
+ * Get the amount that is being transferred
+ * @param {Number} currentBalance
+ * @returns {Number}
+ */
+function getWalletTransferAmount(currentBalance) {
```
Comment (marcaaron) [inline src/libs/PaymentUtils.js:72]: This function does one thing (performs a calculation) and it doesn't look like a "getter" (i.e. PaymentUtils supplies the value). 

Can we please try to give it a name that describes **what it does** like `subtractWalletTransferBalanceFee(currentBalance)`?

File: src/libs/actions/PaymentMethods.js [hunk 17]
```diff
@@ -95,8 +96,59 @@ function clearDebitCardFormErrorAndSubmit() {
...
+ * @param {'bankAccount'|'debitCard'} paymentMethod.type
+ * @returns {Promise}
+ */
+function transferWalletBalance(paymentMethod) {
+    const parameters = {
+        [paymentMethod.type === CONST.PAYMENT_METHODS.BANK_ACCOUNT ? 'bankAccountID' : 'fundID']: paymentMethod.id,
+    };
+    Onyx.merge(ONYXKEYS.WALLET_TRANSFER, {loading: true});
+
+    return API.TransferWalletBalance(parameters)
```
Comment (marcaaron) [inline src/libs/actions/PaymentMethods.js:112]: Let's remove the `return` as it does not look like anything is using the return value of `transferWalletBalance`?

File: src/pages/settings/Payments/TransferBalancePage.js (hunk 3 above)
Comment (parasharrajat) [reply src/pages/settings/Payments/TransferBalancePage.js]: So we use an Onyx key to store the temporary data used for wallet transfer. And the same key is used to determine the isVisible state of ConfirmModal. 
1. After the transfer is done, we don't need the temporary data and thus we have to set the Onyx key to null to clear it.
2. Confirm Modal is shown to user after API call is made and we need `shouldShowConfirmModal` to open this.
3. If we set the Onyx data to null after API call we won't be able to determine the state for ConfirmModal. 

we need a trigger point to clear the Onyx data after wallet transfer i.e. a place where we can set the Key to null. I think the best place would be after user has confirmed the Transfer.

File: src/libs/actions/PaymentMethods.js [hunk 18]
```diff
@@ -95,8 +96,59 @@ function clearDebitCardFormErrorAndSubmit() {
...
+            Growl.error(Localize.translateLocal('transferAmountPage.failedTransfer'));
+            Onyx.merge(ONYXKEYS.WALLET_TRANSFER, {loading: false});
+        });
+}
+
+/**
+ * Set the necessary data for wallet transfer
+ * @param {String} selectedAccountID
+ */
+function saveWalletTransferAccount(selectedAccountID) {
```
Comment (marcaaron) [inline src/libs/actions/PaymentMethods.js:131]: We are saving the `selectedAccountID` and "resetting" the rest of the data. I would recommend we try to do one of these two things:

1. Rename this so that it is very clear **what it does**
2. Split setting the `selectedAccountID` from the reset action and call two functions instead of one

File: src/libs/actions/PaymentMethods.js [hunk 19]
```diff
@@ -95,8 +96,59 @@ function clearDebitCardFormErrorAndSubmit() {
...
+    Onyx.set(ONYXKEYS.WALLET_TRANSFER, {
```
Comment (marcaaron) [inline src/libs/actions/PaymentMethods.js:132]: Is it possible to use a `merge()` here instead of a `set()`?

File: src/libs/actions/PaymentMethods.js [hunk 20]
```diff
@@ -95,8 +96,59 @@ function clearDebitCardFormErrorAndSubmit() {
...
+        selectedAccountID,
+        filterPaymentMethodType: null,
+        loading: false,
+        shouldShowConfirmModal: false,
+    });
+}
+
+/**
+ * Remove all the data related to wallet transfer and close the ConfirmModal
```
Comment (marcaaron) [inline src/libs/actions/PaymentMethods.js:141]: Also think we should not be doing these two things at once. Closing a modal is a very explicit action. Resetting the wallet transfer data is another specific action. Is it easy to tell that calling `clearWalletTransfer()` has a side effect where a modal closes?

File: src/pages/settings/Payments/TransferBalancePage.js [hunk 21]
```diff
@@ -1,27 +1,251 @@
...
+            this.props.cardList,
+        );
+
+        const accountID = this.props.walletTransfer.selectedAccountID || lodashGet(this.props, 'userWallet.walletLinkedAccountID', '');
+        return _.find(paymentMethods, method => method.id === accountID);
+    }
+
+    /**
+     * Transfer Wallet balance
+     * @param {PaymentMethod} selectedAccount
```
Comment (marcaaron) [inline src/pages/settings/Payments/TransferBalancePage.js:120]: `Object`

File: src/libs/actions/PaymentMethods.js (hunk 1 above)
Comment (marcaaron) [reply src/libs/actions/PaymentMethods.js]: I understand that instinct, but my feedback would be that the shorter way is not always the easiest to understand. It's usually better to write code that leaves no doubt about what is happening vs. code where we must inspect the details of every method to figure out what it does.

File: src/libs/PaymentUtils.js (hunk 16 above)
Comment (parasharrajat) [reply src/libs/PaymentUtils.js:72]: Ok. Sure.

File: src/pages/settings/Payments/TransferBalancePage.js (hunk 3 above)
Comment (marcaaron) [reply src/pages/settings/Payments/TransferBalancePage.js]: Sorry, not sure if I totally understand still. But can we solve this problem by using a new key for the modal visibility flag to separate it from the wallet data? That way we can reset or clear the data however we want and have separate controls to show/hide the modal?

Comment (marcaaron) [reply src/pages/settings/Payments/TransferBalancePage.js]: >we have to set the Onyx key to null to clear it

Unsure why we need to set this `null` - what happens if we don't do this?

Comment (parasharrajat) [reply src/pages/settings/Payments/TransferBalancePage.js]: Nothing. It will just remove those keys from storage but I think storage is never a problem for us. So I am correcting it now.

File: src/libs/actions/PaymentMethods.js (hunk 20 above)
Comment (parasharrajat) [reply src/libs/actions/PaymentMethods.js:141]: I agree. Now, I am not worried about the storage so I am refactoring it.

File: src/libs/actions/PaymentMethods.js (hunk 19 above)
Comment (parasharrajat) [reply src/libs/actions/PaymentMethods.js:132]: Yeah. It does not hurt. Sorry, forgot the conflict issues between set and merge.

**Comment by parasharrajat** [comment]
Ready for review.

File: src/styles/utilities/spacing.js [hunk 22]
```diff
@@ -165,6 +165,10 @@ export default {
         marginBottom: -4,
     },
 
+    mrn3: {
+        marginRight: -12,
+    },
```
Comment (marcaaron) [inline src/styles/utilities/spacing.js:170]: NAB, this looks unused, but maybe it will be used in a next PR

**Review by marcaaron — DISMISSED** [review]
Thanks for the changes @parasharrajat! Nice work and thanks for taking the feedback well as always. Haven't tested yet, but code changes look good and this is under a beta so I'm approving. All yours @stitesExpensify.

File: src/styles/utilities/spacing.js (hunk 22 above)
Comment (parasharrajat) [reply src/styles/utilities/spacing.js:170]: Right.

**Comment by parasharrajat** [comment]
@stitesExpensify Bump...

**Comment by stitesExpensify** [comment]
Apologies for the delay, I will review today!

File: src/pages/settings/Payments/TransferBalancePage.js (hunk 2 above)
Comment (stitesExpensify) [reply src/pages/settings/Payments/TransferBalancePage.js]: Yeah we have to set the account manually for now. We should start defaulting them in the future though..

**Comment by stitesExpensify** [comment]
Testing notes:
- If you have no accounts (no card or bank account) the "Which account" is just empty
    - Not sure if this needs to be fixed here since it wasn't directly mentioned in the mockups.

- Wallet balance is actually negative and in cents, so the math is off in the button at the bottom. This is also preventing me from testing the transfer
    -

**Comment by parasharrajat** [comment]
> If you have no accounts (no card or bank account) the "Which account" is just empty

Believe me, we have come far from the mockups. There have been several changes from the initial plan. I can fix this. Just let me know what needs to be done for this. 



> Wallet balance is actually negative and in cents, so the math is off in the button at the bottom. This is also preventing me from testing the transfer

You mean in the calculation, as I see $123.45 as balance. I can quickly test it.

**Comment by parasharrajat** [comment]
@stitesExpensify I fixed the wallet balance calculation. Now it's in dollars and it is a balance - fee. And until you select a default account, the button won't be enabled

**Comment by stitesExpensify** [comment]
> Wallet balance is actually negative and in cents, so the math is off in the button at the bottom. This is also preventing me from testing the transfer

My bad, we actually multiply by -1  before sending it form the server. It is in cents though

**Comment by stitesExpensify** [comment]
"Which Account" also appears to be empty when you _do_ have a default account. Are we just not doing that for this PR?

**Comment by parasharrajat** [comment]
Which account depends on 
>   const accountID = this.props.walletTransfer.selectedAccountID || lodashGet(this.props, 'userWallet.walletLinkedAccountID', '');

So if we have `this.props.walletTransfer.selectedAccountID ` then it will work. But I think this is null. Can you please verify this?

**Comment by stitesExpensify** [comment]
> So if we have this.props.userWallet.walletLinkedAccountID then it will work. But I think this is null. Can you please verify this?

I have an accountID because I have a `this.props.userWallet.walletLinkedAccountID`. It looks like the next line is failing though because the methods don't have IDs there.

**Comment by stitesExpensify** [comment]
[This line](https://github.com/Expensify/App/pull/6878/files#diff-15274a33a8d0582fe2ea47236f1dcc70a2874b5fbd469f9ad07330aac3f748f1R115) could run into problems too if we happen to have multiple methods of different types with the same ID right?

**Comment by stitesExpensify** [comment]
> I fixed the wallet balance calculation. Now it's in dollars and it is a balance - fee. And until you select a default account, the button won't be enabled

Sorry for not being clear, I think we should keep everything in cents (that's what we do on the back end). So instead of dividing by 100, just change the fee constant to cents.

**Comment by parasharrajat** [comment]
> This line could run into problems too if we happen to have multiple methods of different types with the same ID right?

yeah. But I don't think we will ever have the same ids for the two, looking it with the existing payment method list component where we use the id to trigger the click action.

**Comment by parasharrajat** [comment]
Ok, I fixed it @stitesExpensify and changed the amount to cents.

File: src/pages/settings/Payments/PaymentsPage.js [hunk 23]
```diff
@@ -145,6 +180,19 @@ class PaymentsPage extends React.Component {
                         }}
                         onItemSelected={method => this.addPaymentMethodTypePressed(method)}
                     />
+                    <ConfirmModal
```
Comment (stitesExpensify) [inline src/pages/settings/Payments/PaymentsPage.js:183]: Why is this here instead of the transfer page? I must be missing something :smile:

File: src/pages/settings/Payments/TransferBalancePage.js [hunk 24]
```diff
@@ -1,27 +1,251 @@
...
+    constructor(props) {
+        super(props);
+
+        this.paymentTypes = [
+            {
+                key: CONST.WALLET.TRANSFER_METHOD_TYPE.INSTANT,
+                title: this.props.translate('transferAmountPage.instant'),
+                description: this.props.translate('transferAmountPage.instantSummary', {
+                    amount: this.props.numberFormat(
+                        0.25,
```
Comment (stitesExpensify) [inline src/pages/settings/Payments/TransferBalancePage.js:82]: We actually want this to be 1.5% of the total transfer, with a $0.25 minimum

File: src/pages/settings/Payments/PaymentsPage.js (hunk 23 above)
Comment (parasharrajat) [reply src/pages/settings/Payments/PaymentsPage.js:183]: As per the docs, we take the user back to the Payment page and show the confirmation modal for the transfer.

File: src/pages/settings/Payments/TransferBalancePage.js (hunk 24 above)
Comment (parasharrajat) [reply src/pages/settings/Payments/TransferBalancePage.js:82]: Ok. So I will add another method to calculate the real fees. I guess the Fee on the UI will also change.

Comment (stitesExpensify) [reply src/pages/settings/Payments/TransferBalancePage.js:82]: Section 5 in the original issue has said that since the beginning I believe 😉 https://github.com/Expensify/App/issues/3922

File: src/pages/settings/Payments/PaymentsPage.js (hunk 23 above)
Comment (stitesExpensify) [reply src/pages/settings/Payments/PaymentsPage.js:183]: Ah I just looked back at the mockups and see that. Looks good!

File: src/libs/actions/PaymentMethods.js [hunk 25]
```diff
@@ -114,10 +115,68 @@ function clearDebitCardFormErrorAndSubmit() {
...
+        shouldShowConfirmModal: false,
+    });
+}
+
+/**
+ * @param {Number} transferAmount
+ */
+function saveWalletTransferTransferredAmount(transferAmount) {
+    Onyx.merge(ONYXKEYS.WALLET_TRANSFER, {transferAmount});
+}
```
Comment (parasharrajat) [inline src/libs/actions/PaymentMethods.js:163]: I had to store this transferAmount as after user click transfer amount, balance in the userWallet will be set to 0 and then in the ConfirmModal we will not have any way to show the transferred amount.

**Comment by parasharrajat** [comment]
PR Updated. 

I have not added logic to set the Default Account for two reasons:
1. I don't know where to set it. 
2. I don't know How to set it :wink:

**Comment by parasharrajat** [comment]
PR update @marcaaron @stitesExpensify

File: src/libs/PaymentUtils.js [hunk 26]
```diff
@@ -7,7 +7,21 @@ import CONST from '../CONST';
...
+ * Check to see if user has either a debit card or personal bank account added
+ *
+ * @param {Array} [cardList]
+ * @param {Array} [bankAccountList]
+ * @returns {Boolean}
+ */
+function hasExpensifyPaymentMethod(cardList = [], bankAccountList = []) {
+    return _.some(cardList, card => card) || _.some(bankAccountList, (bankAccountJSON) => {
+        const bankAccount = new BankAccount(bankAccountJSON);
+        return bankAccount.isDefaultCredit();
```
Comment (stitesExpensify) [inline src/libs/PaymentUtils.js:19]: Why are we only checking for default credit here? Isn't it any bank account that is verified / open?

Comment (stitesExpensify) [reply src/libs/PaymentUtils.js:19]: cc: @marcaaron

File: src/libs/actions/PaymentMethods.js (hunk 25 above)
Comment (stitesExpensify) [reply src/libs/actions/PaymentMethods.js:164]: I think `saveWalletTransferAmount` is fine descriptiveness-wise, and not as long :smile:

File: src/pages/settings/Payments/TransferBalancePage.js [hunk 27]
```diff
@@ -79,7 +79,7 @@ class TransferBalancePage extends React.Component {
                 title: this.props.translate('transferAmountPage.instant'),
                 description: this.props.translate('transferAmountPage.instantSummary', {
                     amount: this.props.numberFormat(
-                        0.25,
+                        CONST.WALLET.TRANSFER_METHOD_TYPE_FEE.INSTANT.MINIMUM_FEE / 100,
```
Comment (stitesExpensify) [inline src/pages/settings/Payments/TransferBalancePage.js:82]: This needs to be changed to say `1.5% fee ($0.25 minimum)`

Comment (stitesExpensify) [reply src/pages/settings/Payments/TransferBalancePage.js:82]: (and translated)

File: src/pages/settings/Payments/TransferBalancePage.js [hunk 28]
```diff
@@ -129,8 +129,9 @@ class TransferBalancePage extends React.Component {
             ? CONST.WALLET.TRANSFER_METHOD_TYPE.ACH
             : CONST.WALLET.TRANSFER_METHOD_TYPE.INSTANT;
 
-        const transferAmount = PaymentUtils.subtractWalletTransferBalanceFee(this.props.userWallet.currentBalance).toFixed(2);
-        const canTransfer = transferAmount > CONST.WALLET.TRANSFER_BALANCE_FEE;
+        const calculatedFee = PaymentUtils.calculateWalletTransferBalanceFee(this.props.userWallet.currentBalance, selectedPaymentType);
+        const transferAmount = this.props.userWallet.currentBalance - calculatedFee;
+        const canTransfer = transferAmount > calculatedFee;
```
Comment (stitesExpensify) [inline src/pages/settings/Payments/TransferBalancePage.js:134]: Shouldn't this line either be `this.props.userWallet.currentBalance > calculatedFee` or `transferAmount > 0`? 

As it is, you couldn't transfer $0.50 because `50-25 = 25` and `25 !> 25`

**Review by stitesExpensify — COMMENTED** [review]
Looking good, we're getting close!

File: src/libs/PaymentUtils.js (hunk 26 above)
Comment (marcaaron) [reply src/libs/PaymentUtils.js:19]: I might not understand what "default credit" means. Is it not a personal / deposit account?

I thought VBA is not a valid payment method, but maybe wrong about that too. I asked this once [here](https://expensify.slack.com/archives/C01966DBVLL/p1640119438394100?thread_ts=1640119338.393400&cid=C01966DBVLL).

File: src/libs/actions/PaymentMethods.js (hunk 25 above)
Comment (parasharrajat) [reply src/libs/actions/PaymentMethods.js:164]: :eyes:  

:laughing:

File: src/pages/settings/Payments/TransferBalancePage.js (hunk 28 above)
Comment (parasharrajat) [reply src/pages/settings/Payments/TransferBalancePage.js:134]: Yeah, it is supposed to be `transferAmount > 0`. Thanks for checking in.

File: src/libs/actions/PaymentMethods.js [hunk 29]
```diff
@@ -114,10 +116,68 @@ function clearDebitCardFormErrorAndSubmit() {
     });
 }
 
+/**
+ * Call the API to transfer wallet balance.
+ * @param {Object} paymentMethod
+ * @param {String|Number} paymentMethod.methodID
+ * @param {'bankAccount'|'debitCard'} paymentMethod.type
```
Comment (marcaaron) [inline src/libs/actions/PaymentMethods.js:123]: NAB, but these docs are inconsistent with how we do things. If there are multiple types we use `*` and the last one should be `String`. I will try to propose a style guide change so it doesn't need to be mentioned.

File: src/libs/actions/PaymentMethods.js [hunk 30]
```diff
@@ -114,10 +116,68 @@ function clearDebitCardFormErrorAndSubmit() {
...
+ */
+function transferWalletBalance(paymentMethod) {
+    const parameters = {
+        [paymentMethod.type === CONST.PAYMENT_METHODS.BANK_ACCOUNT ? 'bankAccountID' : 'fundID']: paymentMethod.methodID,
```
Comment (marcaaron) [inline src/libs/actions/PaymentMethods.js:127]: Please use the constants for `'bankAccountID'` and `'fundID'` or add them if they don't exist.

File: src/libs/actions/PaymentMethods.js [hunk 31]
```diff
@@ -114,10 +116,68 @@ function clearDebitCardFormErrorAndSubmit() {
...
+    API.TransferWalletBalance(parameters)
+        .then((response) => {
+            if (response.jsonCode !== 200) {
+                throw new Error(response.message);
+            }
+            Onyx.merge(ONYXKEYS.USER_WALLET, {balance: 0});
+            Onyx.merge(ONYXKEYS.WALLET_TRANSFER, {shouldShowConfirmModal: true, loading: false});
+            Navigation.navigate(ROUTES.SETTINGS_PAYMENTS);
+        }).catch((error) => {
+            Log.alert(`[Payments] Failed to transfer wallet balance: ${error.message}`);
```
Comment (marcaaron) [inline src/libs/actions/PaymentMethods.js:140]: Not sure about this alert and think we should remove it for now. We can see in the server logs why a request returned a jsonCode other than 200. @stitesExpensify do you agree?

File: src/libs/actions/PaymentMethods.js [hunk 32]
```diff
@@ -114,10 +116,68 @@ function clearDebitCardFormErrorAndSubmit() {
...
+/**
+ * @param {Number} transferAmount
+ */
+function saveWalletTransferAmount(transferAmount) {
+    Onyx.merge(ONYXKEYS.WALLET_TRANSFER, {transferAmount});
+}
+
+/**
+ * Close the ConfirmModal of Wallet balance transfer
+ */
```
Comment (marcaaron) [inline src/libs/actions/PaymentMethods.js:168]: Remove please.

Future reference: avoid docs that don't add any additional information - https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs

File: src/styles/utilities/spacing.js (hunk 22 above)
Comment (marcaaron) [reply src/styles/utilities/spacing.js:170]: Remove it please.

**Review by marcaaron — CHANGES_REQUESTED** [review]

**Comment by parasharrajat** [comment]
All comments addressed. Ready for final review.

**Review by marcaaron — APPROVED** [review]
Thanks great job!

File: src/pages/settings/Payments/TransferBalancePage.js (hunk 27 above)
Comment (stitesExpensify) [reply src/pages/settings/Payments/TransferBalancePage.js:83]: Bumping this and then we should be good!

**Review by stitesExpensify — COMMENTED** [review]
Almost there!

File: src/pages/settings/Payments/TransferBalancePage.js (hunk 27 above)
Comment (parasharrajat) [reply src/pages/settings/Payments/TransferBalancePage.js:83]: This is being translated correctly as you asked. Its taken rate 1.5% and minimum fee $0.25 from constants.
```
this.props.translate('transferAmountPage.instantSummary', {
                    rate: this.props.numberFormat(CONST.WALLET.TRANSFER_METHOD_TYPE_FEE.INSTANT.RATE),
                    minAmount: this.props.numberFormat(
                        CONST.WALLET.TRANSFER_METHOD_TYPE_FEE.INSTANT.MINIMUM_FEE / 100,
                        {style: 'currency', currency: 'USD'},
                    ),
                }),
 ```     
 ```instantSummary: ({rate, minAmount}) => `${rate}% fee (${minAmount} minimum)```
 
 @stitesExpensify

**Comment by parasharrajat** [comment]
@stitesExpensify Awaiting final review from you and merge.  Hope it answers your question https://github.com/Expensify/App/pull/6878#discussion_r783460445

**Review by stitesExpensify — APPROVED** [review]
Looks great!
